python/cuml/cuml/neighbors/kneighbors_regressor.pyx

```diff
@@ -173,66 +185,120 @@ class KNeighborsRegressor(RegressorMixin,
         Fit a GPU index for k-nearest neighbors regression model.
 
         """
-        if self.weights != "uniform":
-            raise ValueError("Only uniform weighting strategy is supported currently.")
+        if self.weights not in ('uniform', 'distance', None) and not callable(self.weights):
+            raise ValueError(
+                f"weights must be 'uniform', 'distance', or a callable, got {self.weights}"
+            )
 
         super().fit(X, convert_dtype=convert_dtype)
-        self._y = input_to_cuml_array(
+
+        # First get the original dtype before conversion
+        y_temp = input_to_cuml_array(
             y,
             order='F',
-            check_dtype=np.float32,
             check_rows=self.n_samples_fit_,
+            convert_to_dtype=None,  # Don't convert yet, just to get original dtype
+        )
+        self._y_original_dtype = y_temp.dtype
+
+        # Now convert to float32 for internal use
+        y_arr = input_to_cuml_array(
+            y,
+            order='F',
+            check_rows=self.n_samples_fit_,
+            check_dtype=np.float32,
             convert_to_dtype=(np.float32 if convert_dtype else None),
-        ).array
+        )
+        self._y = y_arr.array
+
         return self
 
     @generate_docstring(convert_dtype_cast='np.float32',
                         return_values={'name': 'X_new',
                                        'type': 'dense',
                                        'description': 'Predicted values',
                                        'shape': '(n_samples, n_features)'})
-    @cuml.internals.api_base_return_array(get_output_dtype=True)
+    @cuml.internals.api_base_return_array(get_output_dtype=False)
```

Comment:
If you leave this as `True` you shouldn't need to store `_y_original_dtype` or do any casting internally, the decorator will just handle that output conversion for you. Please use that to avoid the addition of extra state on this estimator.

Reply:
Addressed in [afa5ca3](https://github.com/rapidsai/cuml/pull/7289/commits/afa5ca3dec4fceb6be6a5029b8eb60016d4601d7)